### PR TITLE
Mon 34326 allow to extend engine configuration with a custom file

### PR DESCRIPTION
**Pull request by jean-christophe81**
REFS:MON-34326
## Description

centengine accepts --config-file command line argument to pass a custom json config file

**Fixes** # (issue)

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [X] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 22.04.x
- [ ] 22.10.x
- [ ] 23.04.x
- [ ] 23.10.x
- [X] 24.04.x (master)

<h2> How this pull request can be tested ? </h2>

Please describe the **procedure** to verify that the goal of the PR is matched. Provide clear instructions so that it can be **correctly tested**.

Any **relevant details** of the configuration to perform the test should be added.

## Checklist

- [ ] I have followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [ ] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: engine/src/configuration/extended_conf.cc [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+ *
+ * Centreon Engine is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Centreon Engine. If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
```
Comment (bouda1) [inline engine/src/configuration/extended_conf.cc:18]: Bad header, the licence should be Apache 2.

File: engine/src/configuration/extended_conf.cc [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+#include "com/centreon/exceptions/msg_fmt.hh"
+
+using namespace com::centreon::engine::configuration;
+
+std::list<std::unique_ptr<extended_conf>> extended_conf::_confs;
+
+/**
+ * @brief Construct a new extended state::extended state object
+ *
+ * @param path
```
Comment (bouda1) [inline engine/src/configuration/extended_conf.cc:31]: Fill the param, we don't know if it is a file or a directory

File: engine/src/configuration/extended_conf.cc [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+ * @throw exception if json mal formed
```
Comment (bouda1) [inline engine/src/configuration/extended_conf.cc:32]:
```suggestion
 * @throw exception if json malformed
```

File: engine/src/configuration/extended_conf.cc [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  } catch (const std::exception& e) {
+    SPDLOG_LOGGER_ERROR(
+        log_v2::config(),
+        "extended_conf::extended_conf : fail to read json content from {}: {}",
+        _path, e.what());
+    throw;
+  }
+}
+
+void extended_conf::reload() {
```
Comment (bouda1) [inline engine/src/configuration/extended_conf.cc:51]: An explanation about this function would be interesting. Why would we need to reload() the file and not just load() it ?

File: engine/src/configuration/extended_conf.cc [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    _file_info = file_info;
+  } catch (const std::exception& e) {
+    SPDLOG_LOGGER_ERROR(log_v2::config(),
+                        "extended_conf::extended_conf : fail to read json "
+                        "content from {} => we keep old content, cause: {}",
+                        _path, e.what());
+  }
+}
+
+void extended_conf::apply_all_to_state(state& dest) {
```
Comment (bouda1) [inline engine/src/configuration/extended_conf.cc:73]: Again, an explanation about this function. Do you want to apply the extended_conf to the dest state? Is there another way to apply the extended_conf?
If not, maybe this function could be named `apply(dest)`, like many others in the code...

Comment (bouda1) [reply engine/src/configuration/extended_conf.cc:73]: In fact, `_all_` is often a bad key word. We start with *all* and one day there are exceptions...
And a methode `apply()` would apply also all the conf

File: engine/src/events/loop.cc [hunk 6]
```diff
@@ -101,6 +102,7 @@ static void apply_conf(std::atomic<bool>* reloading) {
       std::string path(::config->cfg_main());
       p.parse(path, config);
     }
+    configuration::extended_conf::apply_all_to_state(config);
     configuration::applier::state::instance().apply(config);
```
Comment (bouda1) [inline engine/src/events/loop.cc:106]: I think I see here why the name.
In fact, you want to update *config* with the extended_conf. Why not calling it 'update_state(...)` in that case?

File: engine/src/main.cc [hunk 7]
```diff
@@ -154,6 +159,10 @@ int main(int argc, char* argv[]) {
         case 'D':  // Diagnostic.
           diagnose = true;
           break;
+        case 'c':
+          if (optarg)
+            extended_conf_path.insert(optarg);
```
Comment (bouda1) [inline engine/src/main.cc:164]: Could we call that variable extended_conf_files as it is a file? Or I missed something...

File: engine/src/main.cc [hunk 8]
```diff
@@ -351,6 +360,10 @@ int main(int argc, char* argv[]) {
           p.parse(config_file, config);
         }
 
+        configuration::extended_conf::load_all(extended_conf_path.begin(),
+                                               extended_conf_path.end());
+
+        configuration::extended_conf::apply_all_to_state(config);
```
Comment (bouda1) [inline engine/src/main.cc:366]: I confirm, I would prefer `update_state(config)`

File: engine/src/configuration/extended_conf.cc (hunk 5 above)
Comment (bouda1) [reply engine/src/configuration/extended_conf.cc:73]: After reading next files, I think something like `update_state(...)` would be better.

File: tests/broker-engine/notification.robot [hunk 9]
```diff
@@ -35,7 +35,7 @@ not1
     Should Be True    ${result}    check_for_external_commands() should be available.
 
     ## Time to set the service to CRITICAL HARD.
-    FOR   ${i}    IN RANGE    ${3}
+    FOR    ${i}    IN RANGE    ${3}
         Ctn Process Service Check Result    host_1    service_1    2    critical
         Sleep    1s
     END
```
Comment (bouda1) [inline tests/broker-engine/notification.robot:41]: We have a keyword now to set this directly without using the loop

File: tests/broker-engine/notification.robot [hunk 10]
```diff
@@ -460,6 +462,7 @@ not10
     Ctn Engine Config Set Value In Hosts    0    host_1    contacts    John_Doe
     Ctn Engine Config Set Value In Contacts    0    John_Doe    host_notification_commands    command_notif
     Ctn Engine Config Set Value In Contacts    0    John_Doe    service_notification_commands    command_notif
+    Ctn ConfigHost Command Status    0    checkh1    1
```
Comment (bouda1) [inline tests/broker-engine/notification.robot:465]: The coding style is not respected.

File: tests/broker-engine/notification.robot [hunk 11]
```diff
@@ -596,7 +596,7 @@ not12
 
     ${content}    Create List    SERVICE ALERT: host_1;service_1;CRITICAL;SOFT;1;critical
     ${result}    Ctn Find In Log With Timeout    ${engineLog0}    ${start}    ${content}    60
-    Should Be True    ${result}    The first service alert SOFT1 is not sent 
+    Should Be True    ${result}    The first service alert SOFT1 is not sent
```
Comment (bouda1) [inline tests/broker-engine/notification.robot:599]:
```suggestion
    Should Be True    ${result}    No log about service_1 in SOFT CRITICAL state.```

**Review by bouda1 — CHANGES_REQUESTED** [review]

File: engine/inc/com/centreon/engine/configuration/extended_conf.hh [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+ * @param end
+ */
+template <class file_path_iterator>
+void extended_conf::load_all(file_path_iterator begin, file_path_iterator end) {
+  _confs.clear();
+  for (; begin != end; ++begin) {
+    try {
+      _confs.emplace_back(std::make_unique<extended_conf>(*begin));
+    } catch (const std::exception&) {
+    }
```
[the remark this answers is not in the document]
Comment (jean-christophe81) [reply engine/inc/com/centreon/engine/configuration/extended_conf.hh:69]: These configurations files are optionals, so I prefer that engine starts even if optional files are not correct

[the remark this answers is not in the document]
Comment (bouda1) [reply engine/inc/com/centreon/engine/configuration/extended_conf.hh:68]: Totally agree, but an error message in the catch to know what happens, even an info message

**Review by bouda1 — APPROVED** [review]

**Pull request merged**